File: crates/defi-protocols/src/lending/euler_v2.rs
use alloy::primitives::{Address, U256};
use alloy::providers::ProviderBuilder;
use alloy::sol;
use alloy::sol_types::SolCall;
use async_trait::async_trait;

use defi_core::error::{DefiError, Result};
use defi_core::traits::Lending;
use defi_core::types::*;

sol! {
    #[sol(rpc)]
    interface IEulerVault {
        function deposit(uint256 amount, address receiver) external returns (uint256);
        function withdraw(uint256 amount, address receiver, address owner) external returns (uint256);
        function borrow(uint256 amount, address receiver) external returns (uint256);
        function repay(uint256 amount, address receiver) external returns (uint256);
        function totalSupply() external view returns (uint256);
        function totalBorrows() external view returns (uint256);
        function interestRate() external view returns (uint256);
    }
}

pub struct EulerV2 {
    name: String,
    euler: Address,
    rpc_url: Option<String>,
}

impl EulerV2 {
    pub fn new(name: String, euler: Address, rpc_url: Option<String>) -> Self {
        Self {
            name,
            euler,
            rpc_url,
        }
    }

    pub fn from_contracts(
        name: String,
        contracts: &std::collections::HashMap<String, Address>,
        rpc_url: Option<String>,
    ) -> Result<Self> {
        let euler = contracts
            .get("euler")
            .or_else(|| contracts.get("markets"))
            .copied()
            .ok_or_else(|| {
                DefiError::ContractError(
                    "Missing 'euler' or 'markets' contract address".to_string(),
                )
            })?;
        Ok(Self::new(name, euler, rpc_url))
    }

    fn rpc_url(&self) -> Result<url::Url> {
        self.rpc_url
            .as_ref()
            .ok_or_else(|| DefiError::RpcError("No RPC URL configured".to_string()))?
            .parse()
            .map_err(|e| DefiError::RpcError(format!("Invalid RPC URL: {e}")))
    }
}

#[async_trait]
impl Lending for EulerV2 {
    fn name(&self) -> &str {
        &self.name
    }

    async fn build_supply(&self, params: SupplyParams) -> Result<DeFiTx> {
        let call = IEulerVault::depositCall {
            amount: params.amount,
            receiver: params.on_behalf_of,
        };

        Ok(DeFiTx {
            description: format!("[{}] Deposit {} into Euler vault", self.name, params.amount),
            to: self.euler,
            data: call.abi_encode().into(),
            value: U256::ZERO,
            gas_estimate: Some(250_000),
        })
    }

    async fn build_borrow(&self, params: BorrowParams) -> Result<DeFiTx> {
        let call = IEulerVault::borrowCall {
            amount: params.amount,
            receiver: params.on_behalf_of,
        };

        Ok(DeFiTx {
            description: format!("[{}] Borrow {} from Euler vault", self.name, params.amount),
            to: self.euler,
            data: call.abi_encode().into(),
            value: U256::ZERO,
            gas_estimate: Some(300_000),
        })
    }

    async fn build_repay(&self, params: RepayParams) -> Result<DeFiTx> {
        let call = IEulerVault::repayCall {
            amount: params.amount,
            receiver: params.on_behalf_of,
        };

        Ok(DeFiTx {
            description: format!("[{}] Repay {} to Euler vault", self.name, params.amount),
            to: self.euler,
            data: call.abi_encode().into(),
            value: U256::ZERO,
            gas_estimate: Some(250_000),
        })
    }

    async fn build_withdraw(&self, params: WithdrawParams) -> Result<DeFiTx> {
        let call = IEulerVault::withdrawCall {
            amount: params.amount,
            receiver: params.to,
            owner: params.to,
        };

        Ok(DeFiTx {
            description: format!(
                "[{}] Withdraw {} from Euler vault",
                self.name, params.amount
            ),
            to: self.euler,
            data: call.abi_encode().into(),
            value: U256::ZERO,
            gas_estimate: Some(250_000),
        })
    }

    async fn get_rates(&self, _asset: Address) -> Result<LendingRates> {
        let url = self.rpc_url()?;
        let provider = ProviderBuilder::new().connect_http(url);
        let vault = IEulerVault::new(self.euler, &provider);

        let total_supply =
            vault.totalSupply().call().await.map_err(|e| {
                DefiError::RpcError(format!("[{}] totalSupply failed: {e}", self.name))
            })?;
        let total_borrows = vault.totalBorrows().call().await.map_err(|e| {
            DefiError::RpcError(format!("[{}] totalBorrows failed: {e}", self.name))
        })?;
        let interest_rate = vault.interestRate().call().await.map_err(|e| {
            DefiError::RpcError(format!("[{}] interestRate failed: {e}", self.name))
        })?;

        // Euler V2 interest rates are per-second rates scaled by 1e27
        let rate_f64 = interest_rate.to::<u128>() as f64 / 1e27;
        let seconds_per_year = 365.25 * 24.0 * 3600.0;
        let borrow_apy = rate_f64 * seconds_per_year * 100.0;

        // Calculate utilization
        let supply_f64 = total_supply.to::<u128>() as f64;
        let borrows_f64 = total_borrows.to::<u128>() as f64;
        let utilization = if supply_f64 > 0.0 {
            borrows_f64 / supply_f64 * 100.0
        } else {
            0.0
        };

        // Supply APY is approximately borrow_apy * utilization_ratio
        let supply_apy = borrow_apy * (borrows_f64 / supply_f64.max(1.0));

        Ok(LendingRates {
            protocol: self.name.clone(),
            asset: _asset,
            supply_apy,
            borrow_variable_apy: borrow_apy,
            borrow_stable_apy: None,
            utilization,
            total_supply,
            total_borrow: total_borrows,
        })
    }

    async fn get_user_position(&self, _user: Address) -> Result<UserPosition> {
        let _url = self.rpc_url()?;

        // Euler V2 user positions are per-vault. A full position query would require
        // iterating over all vaults the user has interacted with.
        // The vault contract address (self.euler) represents a single vault.
        Err(DefiError::Unsupported(format!(
            "[{}] Euler V2 user positions require querying individual vault balances. \
             Use the vault address directly to check balanceOf(user) for supply positions.",
            self.name
        )))
    }
}

File: ts/packages/defi-protocols/src/lending/euler_v2.ts
import { createPublicClient, http, parseAbi, encodeFunctionData } from "viem";
import type { Address } from "viem";
import type { ILending } from "@hypurrquant/defi-core";
import {
  DefiError,
  type ProtocolEntry,
  type SupplyParams,
  type BorrowParams,
  type RepayParams,
  type WithdrawParams,
  type LendingRates,
  type UserPosition,
  type DeFiTx,
} from "@hypurrquant/defi-core";

const EULER_VAULT_ABI = parseAbi([
  "function deposit(uint256 amount, address receiver) external returns (uint256)",
  "function withdraw(uint256 amount, address receiver, address owner) external returns (uint256)",
  "function borrow(uint256 amount, address receiver) external returns (uint256)",
  "function repay(uint256 amount, address receiver) external returns (uint256)",
  "function totalSupply() external view returns (uint256)",
  "function totalBorrows() external view returns (uint256)",
  "function interestRate() external view returns (uint256)",
]);

const SECONDS_PER_YEAR = 365.25 * 24 * 3600;

export class EulerV2Adapter implements ILending {
  private readonly protocolName: string;
  private readonly euler: Address;
  private readonly rpcUrl?: string;

  constructor(entry: ProtocolEntry, rpcUrl?: string) {
    this.protocolName = entry.name;
    this.rpcUrl = rpcUrl;
    const contracts = entry.contracts ?? {};
    const euler = contracts["evk_vault"] ?? contracts["euler"] ?? contracts["markets"];
    if (!euler) throw DefiError.contractError("Missing 'evk_vault' or 'euler' contract address");
    this.euler = euler;
  }

  name(): string {
    return this.protocolName;
  }

  async buildSupply(params: SupplyParams): Promise<DeFiTx> {
    const data = encodeFunctionData({
      abi: EULER_VAULT_ABI,
      functionName: "deposit",
      args: [params.amount, params.on_behalf_of],
    });
    return {
      description: `[${this.protocolName}] Deposit ${params.amount} into Euler vault`,
      to: this.euler,
      data,
      value: 0n,
      gas_estimate: 250_000,
    };
  }

  async buildBorrow(params: BorrowParams): Promise<DeFiTx> {
    const data = encodeFunctionData({
      abi: EULER_VAULT_ABI,
      functionName: "borrow",
      args: [params.amount, params.on_behalf_of],
    });
    return {
      description: `[${this.protocolName}] Borrow ${params.amount} from Euler vault`,
      to: this.euler,
      data,
      value: 0n,
      gas_estimate: 300_000,
    };
  }

  async buildRepay(params: RepayParams): Promise<DeFiTx> {
    const data = encodeFunctionData({
      abi: EULER_VAULT_ABI,
      functionName: "repay",
      args: [params.amount, params.on_behalf_of],
    });
    return {
      description: `[${this.protocolName}] Repay ${params.amount} to Euler vault`,
      to: this.euler,
      data,
      value: 0n,
      gas_estimate: 250_000,
    };
  }

  async buildWithdraw(params: WithdrawParams): Promise<DeFiTx> {
    const data = encodeFunctionData({
      abi: EULER_VAULT_ABI,
      functionName: "withdraw",
      args: [params.amount, params.to, params.to],
    });
    return {
      description: `[${this.protocolName}] Withdraw ${params.amount} from Euler vault`,
      to: this.euler,
      data,
      value: 0n,
      gas_estimate: 250_000,
    };
  }

  async getRates(asset: Address): Promise<LendingRates> {
    if (!this.rpcUrl) throw DefiError.rpcError("No RPC URL configured");
    const client = createPublicClient({ transport: http(this.rpcUrl) });

    const [totalSupply, totalBorrows, interestRate] = await Promise.all([
      client.readContract({ address: this.euler, abi: EULER_VAULT_ABI, functionName: "totalSupply" }).catch((e: unknown) => { throw DefiError.rpcError(`[${this.protocolName}] totalSupply failed: ${e}`); }),
      client.readContract({ address: this.euler, abi: EULER_VAULT_ABI, functionName: "totalBorrows" }).catch((e: unknown) => { throw DefiError.rpcError(`[${this.protocolName}] totalBorrows failed: ${e}`); }),
      client.readContract({ address: this.euler, abi: EULER_VAULT_ABI, functionName: "interestRate" }).catch((e: unknown) => { throw DefiError.rpcError(`[${this.protocolName}] interestRate failed: ${e}`); }),
    ]);

    // Euler V2 interest rates per-second scaled by 1e27
    const rateF64 = Number(interestRate) / 1e27;
    const borrowApy = rateF64 * SECONDS_PER_YEAR * 100;

    const supplyF = Number(totalSupply);
    const borrowF = Number(totalBorrows);
    const utilization = supplyF > 0 ? (borrowF / supplyF) * 100 : 0;
    const supplyApy = borrowApy * (borrowF / Math.max(supplyF, 1));

    return {
      protocol: this.protocolName,
      asset,
      supply_apy: supplyApy,
      borrow_variable_apy: borrowApy,
      utilization,
      total_supply: totalSupply as bigint,
      total_borrow: totalBorrows as bigint,
    };
  }

  async getUserPosition(_user: Address): Promise<UserPosition> {
    throw DefiError.unsupported(
      `[${this.protocolName}] Euler V2 user positions require querying individual vault balances. Use the vault address directly to check balanceOf(user) for supply positions.`,
    );
  }
}
